DAQ/src/StrawAndCaloDigisFromFragments_module.cc

```diff
@@ -70,37 +70,54 @@ StrawAndCaloDigisFromFragments::StrawAndCaloDigisFromFragments(fhicl::ParameterS
   , caloFragmentsTag_(pset.get<art::InputTag>("caloTag","daq:calo"))
 {
   produces<EventNumber_t>(); 
-  produces<mu2e::StrawDigiCollection>();
-  produces<mu2e::CaloDigiCollection>();
+  if (parseTRK_){
+    produces<mu2e::StrawDigiCollection>();
+  }
+  if (parseCAL_){
+    produces<mu2e::CaloDigiCollection>();
+  }
 }
 
 // ----------------------------------------------------------------------
 
 void
-  StrawAndCaloDigisFromFragments::
-  produce( Event & event )
+StrawAndCaloDigisFromFragments::
+produce( Event & event )
 {
-
   art::EventNumber_t eventNumber = event.event();
 
-  auto trkFragments = event.getValidHandle<artdaq::Fragments>(trkFragmentsTag_);
-  auto calFragments = event.getValidHandle<artdaq::Fragments>(caloFragmentsTag_);
-  size_t numTrkFrags = trkFragments->size();
-  size_t numCalFrags = calFragments->size();
+  art::Handle<artdaq::Fragments> trkFragments, calFragments;
+  size_t numTrkFrags(0), numCalFrags(0);
+  if (parseTRK_){
+    event.getByLabel(trkFragmentsTag_ , trkFragments);
+    if (!trkFragments.isValid()){            
+      return;
```

Comment:
We need to have a policy discussions that involve both online and offline.

Suppose that we have an event in which the tracker is present but the calorimeter is not.  Or vice versa. That might often happen during commissioning and might happen from time to time during physics running when an error happens.  Do you really want to skip the good data just because something else is missing?  It's not useful for physics but it might very well be important for debugging.

The second policy is issue is that this code declares that it will produce something and then never produces it. I believe that for Offline we currently have art configured so that it's an error that art will catch and throw an exception.  There is run-time configuration to allow failure to put to be allowed.  I believe that there is also something like produce_maybe which allows us exempt single modules from this policy.

I am not asking for an action at this time but we need to discuss this the online team.